redash_toolbelt/examples/migrate.py

```diff
@@ -0,0 +1,313 @@
+                if str(v['id']) in meta['visualizations']:
+                    continue
+
+                user = user_with_api_key(query['user']['id'])
+                data = {
+                    "name": v['name'],
+                    "description": v['description'],
+                    "options": v['options'],
+                    "type": v['type'],
+                    "query_id": new_query_id
+                }
+                response = requests.post(
+                    DESTINATION + '/api/visualizations', json=data, headers=auth_headers(user['api_key']))
+                response.raise_for_status()
+
+                meta['visualizations'][v['id']] = response.json()['id']
+
+
+def import_dashboards():
+    print("Importing dashboards...")
+
+    dashboards = get_paginated_resource(ORIGIN + '/api/dashboards', ORIGIN_API_KEY)
+
+    for dashboard in dashboards:
+        print("   importing: {}".format(dashboard['slug']))
+        d = api_request('/api/dashboards/{}'.format(dashboard['slug']))
+        data = {'name': d['name']}
+        user = user_with_api_key(d['user_id'])
+        response = requests.post(
+            DESTINATION + '/api/dashboards', json=data, headers=auth_headers(user['api_key']))
+        response.raise_for_status()
+
+        new_dashboard = response.json()
+        requests.post(
+            DESTINATION + '/api/dashboards/{}'.format(new_dashboard['id']), json={'is_draft': False}, headers=auth_headers(user['api_key']))
+
+        # recreate widget
+        for widget in d['widgets']:
+            data = {
+                'dashboard_id': new_dashboard['id'],
+                'options': widget['options'],
+                'width': widget['width'],
+                'text': widget['text'],
+                'visualization_id': None
+            }
+
+            if not isinstance(widget['options'], dict):
+                widget['options'] = {}
+
+            if 'visualization' in widget:
+                data['visualization_id'] = meta['visualizations'].get(
+                    str(widget['visualization']['id']))
+
+            if 'visualization' in widget and not data['visualization_id']:
+                print('skipping for missing viz')
+                continue
+
+            response = requests.post(
+                DESTINATION + '/api/widgets', json=data, headers=auth_headers(user['api_key']))
+            response.raise_for_status()
+
+
+def fix_queries():
+    """
+    This runs after importing all queries, so we can update the query id reference
+    in parameter definitions.
+    """
+    print("Updating queries options...")
+
+    for query_id, new_query_id in meta['queries'].items():
+        query = api_request('/api/queries/{}'.format(query_id))
+        print("   Fixing: {}".format(query_id))
+
+        options = query['options']
+        for p in options.get('parameters', []):
+            if 'queryId' in p:
+                p['queryId'] = meta['queries'].get(str(p['queryId']))
```

Comment:
This might fail if the query has no values. Worth triggering a refresh, wait for it to complete (with some timeout?) and then continue.

Reply:
I hesitate to kick this off synchronously since we a) don't want to hammer databases with queries b) delay the rest of the migration if these queries are long-running.